engine: add 23.2 beta (#1390)
redpanda-data/documentation: modules/manage/pages/use-admin-api.adoc

= Manage Redpanda using the Admin API
:page-categories: Management, High Availability, Upgrades
:description: Manage components of a Redpanda cluster, such as individual brokers and partition leadership. The Redpanda Admin API also allows you to perform operations that are specific to Redpanda Self-Managed and cannot be done using the standard Kafka API.

The Redpanda Admin API allows you to manage your cluster and perform operations specific to Redpanda Self-Managed that are not available through the standard Kafka API. You can call the Admin API using any HTTP client. 

Most Admin API operations are also available using xref:get-started:intro-to-rpk.adoc[`rpk`], a CLI tool that interacts with the Admin API under the hood. 

> Redpanda v25.3 introduces new endpoints to the Admin API that are served with https://connectrpc.com/docs/introduction[ConnectRPC]. New Redpanda features and operations available starting in v25.3 are accessible as RPC services through these new endpoints. Existing Admin API operations from versions earlier than 25.3 remain available at the same URLs and you can continue to use them as usual (including with rpk v25.3 and later).

Reviewer: ...remain available at the same URLs...
The same URLs as what?

== Prerequisites

* A running Redpanda Self-Managed cluster.

> * Superuser privileges, if xref:manage:security/authentication.adoc#enable-authentication[authentication] is enabled on your cluster for the Admin API. For more information, see xref:manage:security/authentication.adoc#create-superusers[Configure Authentication]. (Some endpoints are read-only and do not require superuser access.)

Reviewer: So if I want to know if the endpoints I need to use require superuser access, can I view a list somewhere that tells me which endpoints require it and which ones don't?

Reviewer 2: All v2 endpoints require superuser, but I don't have a definitive list for v1. Will reach out to an SME - if this list can be put together, these details can be added at a later time

* A tool to make HTTP requests, such as `curl`, or client libraries for your programming language of choice. 

> ** For Admin API operations introduced in v25.3 and later, you can also make requests using a ConnectRPC client. You can install the Connect plugin for your preferred language and use the Protobuf compiler to generate an SDK. These RPC services are also available in a Buf module and can be accessed through the https://buf.build/redpandadata/core/docs/dev:redpanda.core.admin.v2[Buf Schema Registry]. The https://buf.build/docs/cli/[Buf CLI] provides an easy way to generate client SDKs.

Reviewer: ```suggestion
** For Admin API operations introduced in v25.3 and later, you can also make requests using a ConnectRPC client. You can install the Connect plugin for your preferred language and use the Protobuf compiler to generate an SDK. These RPC services are also available in a Buf module, which you can access through the https://buf.build/redpandadata/core/docs/dev:redpanda.core.admin.v2[Buf Schema Registry]. The https://buf.build/docs/cli/[Buf CLI] provides an easy way to generate client SDKs.
```

== Use the Admin API

Starting in Redpanda v25.3, the Admin API serves new endpoints as ConnectRPC services, which allows you to continue to invoke them using HTTP requests. Both new and legacy (RESTful) endpoints are accessible on the same port (default: 9644), but they use different URL paths.

NOTE: Legacy Admin API endpoints remain available and fully supported. Use them for operations not yet available as ConnectRPC services.

=== Authentication

If authentication is enabled on your cluster, you must provide credentials with each request, either using HTTP Basic authentication or by including an `Authorization` header with a bearer token. For example:

[,bash]
----
curl -u <user>:<password> -X GET "http://localhost:9644/v1/cluster_config"
----

=== Use legacy (RESTful) endpoints

The base URL for all requests to the legacy endpoints is:

```
http://<broker-address>:<admin-api-port>/v1/
```

// TODO: Update link if necessary when v2 URLs are finalized

> For a full list of available endpoints, see the link:/api/doc/admin/[Admin API Reference]. Select "v1" in the version selector to view legacy endpoints.

Reviewer: link to the Admin API Reference does not render in html output - throws a Page Not Found

Reviewer 2: This is a limitation with our Bump-hosted (`link:/`) pages, they don't get included in the preview build

==== Example request

To use the Admin API to xref:manage:cluster-maintenance/decommission-brokers.adoc[decommission a broker]:

[tabs]
====
curl::
+
--
Send a PUT request to the link:/api/doc/admin/operation/operation-decommission[`/v1/brokers/\{broker_id}/decommission`] endpoint:

[,bash]
----
curl \
 -u <user>:<password> \
 --request PUT 'http://<broker-address>:<port>/v1/brokers/<broker-id>/decommission'
----
--

rpk::
+
--
For Linux deployments only, run xref:reference:rpk/rpk-redpanda/rpk-redpanda-admin-brokers-decommission.adoc[`rpk redpanda admin brokers decommission`]:

[,bash]
----
rpk redpanda admin brokers decommission <broker-id>
----
--
====

> === Use new (ConnectRPC) endpoints

Reviewer: Maybe consider: Use ConnectRPC endpoints

and then put a note explaining what makes them different from v1 endpoints.

> You can send `curl` requests with a JSON payload, or use a generated ConnectRPC client to call the same methods directly from your application code. The URL paths are the fully qualified names of the ConnectRPC services. These endpoints accept POST requests only.

Reviewer: ```suggestion
You can send `curl` requests with a JSON payload, or use a generated ConnectRPC client to call the same methods directly from your application code. The URL paths are the fully-qualified names of the ConnectRPC services. These endpoints accept POST requests only.
```

> Use the new ConnectRPC endpoints with features such as:

Reviewer: ```suggestion
Use ConnectRPC endpoints with features such as:
```

// TODO: Add links to docs when they are merged
* Shadowing
* Connected client monitoring

> For a full list of available endpoints, see the link:/api/doc/admin/v2/[Admin API Reference]. Select "v2" in the version selector to view the ConnectRPC endpoints.

Reviewer: @mattschumpert I think users are going to be throwing questions our way--we should be up front/transparent about v1 vs v2. If we are not, this will generate questions and may undermine user confidence in the docs.

= Manage Redpanda using the Admin API
:page-categories: Management, High Availability, Upgrades
:description: Manage components of a Redpanda cluster, such as individual brokers and partition leadership. The Redpanda Admin API also allows you to perform operations that are specific to Redpanda Self-Managed and cannot be done using the standard Kafka API.

The Redpanda Admin API allows you to manage your cluster and perform operations specific to Redpanda Self-Managed that are not available through the standard Kafka API. You can call the Admin API using any HTTP client. 

> Most Admin API operations are also available using xref:get-started:intro-to-rpk.adoc[`rpk`], a CLI tool which interacts with the Admin API under the hood. 

Reviewer: ```suggestion
Most Admin API operations are also available using xref:get-started:intro-to-rpk.adoc[`rpk`], a CLI tool that interacts with the Admin API under the hood. 
```

> Redpanda version 25.3 introduces a version 2 of the Admin API, powered by https://connectrpc.com/docs/introduction[ConnectRPC]. New Redpanda features and operations introduced starting in version 25.3 are available with Admin API v2 only, and are not backported to Admin API v1. Supported Admin API v1 operations remain available and you can continue to use them per usual (including with rpk v25.3 and later).

Reviewer: I keep seeing v25.3 in other contexts. No big deal, but we should all be on the same page wrt how we do this.

> Redpanda version 25.3 introduces a version 2 of the Admin API, powered by https://connectrpc.com/docs/introduction[ConnectRPC]. New Redpanda features and operations introduced starting in version 25.3 are available with Admin API v2 only, and are not backported to Admin API v1. Supported Admin API v1 operations remain available and you can continue to use them per usual (including with rpk v25.3 and later).

Reviewer: ```suggestion
Redpanda v25.3 introduces a version 2 of the Admin API, powered by https://connectrpc.com/docs/introduction[ConnectRPC]. New Redpanda features and operations introduced in version 25.3 are available with Admin API v2 only, and are not backported to Admin API v1. Supported Admin API v1 operations remain available and you can continue to use them as usual (including with rpk v25.3 and later).
```

== Prerequisites

* A running Redpanda Self-Managed cluster.
** For Admin API v1, any supported version of Redpanda. For Admin API v2, Redpanda version 25.3 or later.
* Superuser privileges, if xref:manage:security/authentication.adoc#enable-authentication[authentication] is enabled on your cluster for the Admin API. For more information, see xref:manage:security/authentication.adoc#create-superusers[Configure Authentication]. (Some endpoints are read-only and do not require superuser access.)
* A tool to make HTTP requests, such as `curl`, or client libraries for your programming language of choice. 

> ** For Admin API v2, you can also make requests using a Connect client. You can install the Connect plugin for your preferred language and use the Protobuf compiler to generate an SDK. Admin API v2 is also available as a Buf module and can be accessed through the https://buf.build/redpandadata/core/docs/dev:redpanda.core.admin.v2[Buf Schema Registry]. The https://buf.build/docs/cli/[Buf CLI] provides an easy way to generate client SDKs.

Reviewer: Is this a Kafka Connect? If you are referring generally to connectors, wouldn't this be "connect"?
Or is this Redpanda Connect?

> ** For Admin API v2, you can also make requests using a Connect client. You can install the Connect plugin for your preferred language and use the Protobuf compiler to generate an SDK. Admin API v2 is also available as a Buf module and can be accessed through the https://buf.build/redpandadata/core/docs/dev:redpanda.core.admin.v2[Buf Schema Registry]. The https://buf.build/docs/cli/[Buf CLI] provides an easy way to generate client SDKs.

Reviewer: ```suggestion
** For Admin API v2, you can also make requests using a Connect client. You can install the Connect plugin for your preferred language and use the Protobuf compiler to generate an SDK. Admin API v2 is also available as a Buf module, which you can access using the https://buf.build/redpandadata/core/docs/dev:redpanda.core.admin.v2[Buf Schema Registry]. The https://buf.build/docs/cli/[Buf CLI] provides an easy way to generate client SDKs.
```

== Choose an Admin API version

> In Redpanda version 25.3, Admin API v2 endpoints are entirely exclusive to v2, and there is virtually no overlap in functionality between v2 and v1. You can use both versions of the Admin API on the same cluster, so choosing which version depends on whether you use the following v2-only features:

Reviewer: I would not mix v and version, especially in the same paragraph

> In Redpanda version 25.3, Admin API v2 endpoints are entirely exclusive to v2, and there is virtually no overlap in functionality between v2 and v1. You can use both versions of the Admin API on the same cluster, so choosing which version depends on whether you use the following v2-only features:

Reviewer: ```suggestion
In Redpanda v25.3, Admin API v2 endpoints are entirely exclusive to v2, and there is virtually no overlap in functionality between v2 and v1. You can use both versions of the Admin API on the same cluster, so choosing which version depends on whether you use the following v2-only features:
```

[,bash]
----
curl -u <user>:<password> -X GET "http://localhost:9644/v1/cluster_config"
----

=== Use Admin API v1

The Admin API v1 uses RESTful HTTP endpoints. The base URL for all requests is:

```
http://<broker-address>:<admin-api-port>/v1/
```

For a full list of available endpoints, see the link:/api/doc/admin/[Admin API v1 Reference].

==== Example request

To use the Admin API v1 to xref:manage:cluster-maintenance/decommission-brokers.adoc[decommission a broker]:

[tabs]
====
curl::
+
--
Send a PUT request to the link:/api/doc/admin/operation/operation-decommission[`/v1/brokers/\{broker_id}/decommission`] endpoint:

[,bash]
----
curl \
 -u <user>:<password> \
 --request PUT 'http://<broker-address>:<port>/v1/brokers/<broker-id>/decommission'
----
--

rpk::
+
--
For Linux deployments only, run xref:reference:rpk/rpk-redpanda/rpk-redpanda-admin-brokers-decommission.adoc[`rpk redpanda admin brokers decommission`]:

[,bash]
----
rpk redpanda admin brokers decommission <broker-id>
----
--
====

=== Use Admin API v2

> The Admin API v2 is built on ConnectRPC, so you can send `curl` requests with a JSON payload, or use a generated Connect client to call the same methods directly from your application code. The v2 paths are the fully qualified names of the services. All v2 endpoints accept POST requests only.

Reviewer: Connect client<--- Redpanda Connect? Kafka Connect? connect client?

For a full list of available endpoints, see the link:/api/doc/admin/v2[Admin API v2 Reference].

==== Limitations

> Admin API v2 does not currently include Admin API v1 functionality. Use v1 for operations not available in v2. Feature parity is planned for future releases. In the meantime, v1 remains fully supported.

Reviewer: ```suggestion
Admin API v2 does not include Admin API v1 functionality. Use v1, which remains fully supported, for operations not available in v2. 
```
